### ISSUE TITLE
SFTP (SSH) backup from Windows to Synology NAS not possible, manual connection works

### ISSUE THREAD

**Issue by Christopher87R** [duplicati/duplicati#5546]
- [ x] I have searched open and closed issues for duplicates.
- [ x] I have searched the [forum](https://forum.duplicati.com) for related topics.

----------------------------------------

## Environment info

- **Duplicati version**: 2.0.8.1_beta_2024-05-07
- **Operating system**: Windows 11 Version 23H2 (Build 22631.4112)
- **Backend**: SFTP Server using Synology NAS

## Description

Unfortunately I cannot establish a connection via SFTP (SSH), I get the following error when testing the connection: “Failed to connect: A connection could not be established because the target computer refused the connection”.

I have also tried it via the command line program:
.\Duplicati.CommandLine.BackendTool.exe LIS T "ssh://SECRET.dynv6.net:9999/Backup-WIN-CR?auth-username=backup&auth-password=ABCDEFG" but get the following error: 
"Command failed: Es konnte keine Verbindung hergestellt werden, da der Zielcomputer die Verbindung verweigerte
System.Net.Sockets.SocketException (0x80004005): Es konnte keine Verbindung hergestellt werden, da der Zielcomputer die Verbindung verweigerte"

The credentials are correct, I have tried sftp and ssh, the latter also from my Windows computer (see screenshots)

## Screenshots
![image](https://github.com/user-attachments/assets/7bea7dd4-ca77-4b7d-83f1-43e2a24ed6e9)

![image](https://github.com/user-attachments/assets/88211344-7882-42cd-88bf-f758591e0040)

![image](https://github.com/user-attachments/assets/1bc15999-4a1a-4657-b49a-d16c98051155)

**Comment by kenkendk** [comment]
The error message indicates that the error happens on the network level.
Not sure how much is example/obfuscated here, but I can see the port is `9999` for the Duplicati command and the screenshot port start with `1`.

If this is just due to obfuscation, it is possible that Duplicati is not trusted by your firewall and this prevents it from making external connections.

**Comment by Christopher87R** [comment]
I used the same ports, this was just for Obfuscation. 
As I also wrote, I can use sftp directly from a Linux computer, ssh also works from Windows and Linux. So the connection works in principle, even the login data is correct.

**Comment by kenkendk** [comment]
I just tested with on my system towards two different SSH servers (one Synology) and the connection works for both.

Since you can get other programs to connect, I think this is a security setup that is preventing Duplicati from communicating with the server.

On Windows, you can add Duplicati to the list of allowed network programs. There is a guide here: https://support.microsoft.com/en-us/windows/risks-of-allowing-apps-through-windows-defender-firewall-654559af-3f54-3dcf-349f-71ccd90bcc5c

Look at the section "Add or remove an app from the list of allowed apps"

**Comment by Christopher87R** [comment]
Thanks, that motivated me again and in the end I found out that the IP (v4) works directly, but not the hostname, although this can be resolved on the Windows and Linux machine.
How does the DNS resolution work in Duplicati?

**Comment by ts678** [comment]
> How does the DNS resolution work in Duplicati?

Host name resolution (which goes beyond DNS) is usually left to the communications library, the C# runtime, and the OS.

https://github.com/duplicati/duplicati/blob/78ea50c8335772f2684ff39f8e44c22d2836322b/Duplicati/Library/Backend/SSHv2/SSHv2Backend.cs#L118

https://github.com/duplicati/duplicati/blob/78ea50c8335772f2684ff39f8e44c22d2836322b/Duplicati/Library/Backend/SSHv2/SSHv2Backend.cs#L317

I don't know the [SSH.NET library code](https://github.com/sshnet/SSH.NET) internals, but I see some use of the following call, and then take the first on the list:

[Dns.GetHostAddresses Method](https://learn.microsoft.com/en-us/dotnet/api/system.net.dns.gethostaddresses?view=netframework-4.8.1) (Microsoft .NET Framework) doesn't specify an order, so make sure all the addresses work.
Troubleshooting might be possible using Wireshark, tcpdump, etc. e.g. if you want to try to find which address was failing.
BTW this message text sounds to me more like TCP RST than a firewall drop. If so, you got to a host but port wasn't active.

**Comment by kenkendk** [comment]
@ts678 I agree this looks like an active close error, as opposed to firewall drop.

If DNS order is weird, you could be hitting another server, but it would be strange if this is consistent and not random.

I found this [other issue that mentions the error code 0x80004005](https://github.com/shadowsocks/shadowsocks-windows/issues/2853) but has a different error message that looks more like a permission issue. 

@Christopher87R, if this only happens with a hostname, it could be "something" is preventing Duplicati from making DNS queries? If this is the case, I would guess the update check would also fail with a similar error message.

**Comment by Christopher87R** [comment]
Sorry for the late reply.

@kenkendk The update query works without any problems. 

My guess is that it is related to IPv6. There is also an AAAA entry for the domain, perhaps the domain is resolved to the IPv6 address?!

When I send a ping in Windows, IPv6 is also used, unless I explicitly specify the use of v4. Could this be a possible cause?

**Comment by ts678** [comment]
> I can confirm that the IPv6 address is preferred

Or if not preferred, at least selected somehow. Having all addresses work still seems like the safest way to go.

> there is no port sharing for it.

[What services on my Synology NAS support IPv6?](https://kb.synology.com/en-sg/DSM/tutorial/What_applications_on_Synology_NAS_are_IPv6_supported) at Synology Knowledge Center seems to say it should work.

**Comment by Christopher87R** [comment]
> [What services on my Synology NAS support IPv6?](https://kb.synology.com/en-sg/DSM/tutorial/What_applications_on_Synology_NAS_are_IPv6_supported) at Synology Knowledge Center seems to say it should work.

Sure, I could probably set up the setup to have full IPv6 support.
In general, unfortunately, IPv6 is still not fully supported (it's a shame). 

I personally would prefer it if both v4 and v6 are attempted.

**Comment by kenkendk** [comment]
@Christopher87R 
> UPDATE: In fact, I can confirm that the IPv6 address is preferred and there is no port sharing for it.

Does this mean that on your system, the SSH backend attempts to connect over IPv6 by default? 
Does this happen because the hostname has the `AAAA` record, so the DNS resolver returns both an IPv4 and IPv6 address, and then SSH (presumably) chooses IPv6?
